File: java/Talk.java
import java.util.regex.*;
import java.util.*;

class Talk {

    //natural
    private String title;
    private int duration;
    public int no;

   // public Track atTrack;
    public Session atSession;

    private Calendar startTime;


    public int movingCount=0; //hanyszor lett ellopva
    public static final int MAX_MOVING_ALLOWED = 5;


    public String getTitle () {
        return title;
    }

    public int getDuration () {
        return duration;
    }


    public static Talk create (String raw) {

        //for a lighning, no need to parse, we know it's 5 mins long
        if (raw.indexOf ("lightning") != -1) {
            return new Talk (raw, 5);
        }

        //build regex to parse a raw line
        String regex = "^(.+) (\\d+)min.*$";
        Pattern r = Pattern.compile (regex);
        Matcher m = r.matcher (raw);

        if (m.find()) {
              return new Talk (m.group(1), Integer.parseInt(m.group(2)));
        } else {
            System.out.println ("ERROR parsing Talk from input line '"+raw+"'. Include mins or lightning keyword (=5mins)");
            return null;
        }
    }

    public void unchain () {
        if (atSession==null) return;
        atSession.talks.remove(this);
        atSession=null;
    }


    public Talk (String t, int d) {
        title = t;
        duration = d;
    }

    public Talk (String t, int d, Session s) {
        title = t;
        duration = d;
        atSession = s;
    }

    public void setStartTime (int offset_min) {
        startTime = (Calendar)atSession.startTime.clone();
        startTime.add (Calendar.MINUTE, offset_min);
    }

    public Calendar getStartTime () {
        return startTime;
    }

    public int getStartTimeOffsetMins () {
        return (int)(startTime.getTimeInMillis () / Workshop.MILLIS_PER_MIN);
    }

    public void printTalk () {
      String ret = "";
      int hr = startTime.get(Calendar.HOUR_OF_DAY);
      int min = startTime.get(Calendar.MINUTE);

      if (startTime != null) ret+= String.format("%02d:%02d", hr, min);
      ret += " "+getTitle();
      if (getDuration()>0) ret+=" (" +getDuration()+" mins)";
      System.out.println (ret);

    }

}

File: scala/example/Talk.scala
package example
import java.util.Calendar
import scala.util.matching.Regex


object Talk {
    val MAX_MOVING_ALLOWED : Int = 5;
    
    def create (raw: String): Talk = {
       
        //for a lighning, no need to parse, we know it's 5 mins long
        if (raw.indexOf ("lightning") != -1) {
            return new Talk (raw, 5, null);
        }
        
        //build regex to parse a raw line
        val patt = """(.+) (\d+)min.*""".r;
        val ret = patt.findFirstMatchIn (raw);
        
        ret match {
            case Some(m) => {
                return new Talk (m.group(1), m.group(2).toInt, null);
            }
            case None => {
                return null;
            }
        }        
  
    return null;       
    }
        
}

class Talk (val t: String, 
            val d: Int,
            val se: Session   
            ) {
    
    val _title : String = t
    val _duration : Int = d
    
    var startTime : Calendar = Calendar.getInstance();
    var _atSession : Session = se;
    
    private var _movingCount =0;
    
    //lightweight getters as methods
    def duration = _duration;
    def title = _title;
    def atSession = _atSession;
    
    def wasMoved = _movingCount += 1;
    def movedTooManyTimes : Boolean = { _movingCount >= Talk.MAX_MOVING_ALLOWED }        
    def wasNeverMoved = { _movingCount = 0; }
    
    //def getMovingCount = _movingCount;
    
    
    def unchain = {
        if (_atSession!=null) {
            _atSession-=this; 
            _atSession=null;
        }
    }
    
    def setSession_= (s: Session) = {
        _atSession = s;
        _atSession += this;
    }
    
    def setStartTime (offset_min: Int) = {                
        startTime = atSession.startTime.clone().asInstanceOf[Calendar]
        startTime.add (Calendar.MINUTE, offset_min)
    }
    
    def getStartTimeOffsetMins : Int = {
        val mins : Long = (startTime.getTimeInMillis () / Workshop.MILLIS_PER_MIN);        
        return mins.toInt;
    }
    
    def printTalk  = {
      var ret : String = "";
      val hr : Int = startTime.get(Calendar.HOUR_OF_DAY);
      val min : Int = startTime.get(Calendar.MINUTE);

      if (startTime != null) ret+= "%02d:%02d".format (hr, min);
      ret += " "+title;
      if (duration>0) ret+=" (" +duration+" mins)";
      System.out.println (ret);

    }

}